synapse/storage/data_stores/main/room.py

```diff
@@ -695,6 +701,69 @@ def _background_insert_retention_txn(txn):
 
         defer.returnValue(batch_size)
 
+    async def _background_add_rooms_room_version_column(
+        self, progress: dict, batch_size: int
+    ):
+        """Background update to go and add room version inforamtion to `rooms`
+        table from `current_state_events` table.
+        """
+
+        last_room_id = progress.get("room_id", "")
+
+        def _background_add_rooms_room_version_column_txn(txn: LoggingTransaction):
+            sql = """
+                SELECT room_id, json FROM current_state_events
```

Comment:
is this going to table-scan `CSE`? shouldn't we base it on `rooms` instead?

Reply:
There's a unique index on `(room_id, type, state_key)` so it just walks the index.

Reply:
right, but it's going to need to walk the *entire* index? Or am I misunderstanding how btree indexes work?

Reply:
Well, it'll iterate over each room and then look up the create event for each within that, AIUI.

The query plan for it on jki.re seems to show it doing that:

```
 Limit  (cost=1.11..12920.09 rows=1 width=1090)
   ->  Nested Loop  (cost=1.11..12920.09 rows=1 width=1090)
         ->  Index Scan using current_state_events_room_id_type_state_key_key on current_state_events  (cost=0.55..12846.65 rows=16 width=64)
               Index Cond: ((room_id > ''::text) AND (type = 'm.room.create'::text) AND (state_key = ''::text))
         ->  Index Scan using event_json_event_id_key on event_json  (cost=0.56..4.58 rows=1 width=1125)
               Index Cond: (event_id = current_state_events.event_id)
               Filter: (current_state_events.room_id = room_id)
```

Reply:
right, my concern is the index scan over that entire index, which is presumably huge on big servers, and it seems like a bunch of unnecessary I/O when we don't care about all the other events in that index.

this isn't a blocking comment; if you think it's going to be ok then that's fine. it was just a thought.

Reply:
I'm reasonably sure it'll be fine. It will efficiently jump straight to the correct room ID, and jumping to the next room ID is fast (it won't have to iterate over all the state in the room, as it uses the tree structure to do it), so it should just be looking up the next room ID in the btree index, using the index to pull out the create event ID, repeating.